### ISSUE TITLE
OpenBrowser throws error with AddExtensions

### ISSUE THREAD

**Issue by dornech** [GCuser99/SeleniumVBA#114]
I use Chrome and AddExtensions with a CRX file downloaded from the chrome webstore. When starting the browser, the error below is shown:
![image](https://github.com/user-attachments/assets/ab95cf34-59b4-4f94-89a0-cf63c8419ec8)

**Comment by GCuser99** [comment]
Hi @dornech,

Any chance you can share minimal reproducible code for this? That would help us debug the issue.

Thanks!

**Comment by dornech** [comment]
Thanks for caring. I am impressed by SeleniumVBA and plan to migrate my applications to SeleniumVBA from SeleniumBasic (which still works somehow but is outdated and ha somne quirks int he API).

```
Set objWebCapabilities = objSelenium.CreateCapabilities
objWebCapabilities.AddExtensions Array("<extension-file including path with extension .crx>")
objWebDriver.StartChrome
objSelenium.OpenBrowser objWebCapabilities

```
I am pretty confident that the code is OK because if I do not do the AddExtensions call, the browser starts and everything works fine.

BTW: Windows10, latest Chrome-Browser

**Comment by GCuser99** [comment]
Thanks for the code - I agree with you - nothing looks unusual to me. BTW, you don't need to wrap your extension path(s) in an Array() structure, just to let you know - you can just list them as comma-separated arguments. I doubt that has anything to do with the issue though... We will get back to you soon.

**Comment by dornech** [comment]
Thanks for the response, yes I I learned already from looking into the code that the array-stuff is not required. Initially thought this to be required from the wiki.

However, strange thing is, that the CRX file works with SeleniumBasic, so it is not an issue of the CRX file but seems to be an issue how the options are prepared for / presented to webdriver / chromnedriver.exe.

**Comment by GCuser99** [comment]
Ok, I see the direction that we have to go to fix this: https://stackoverflow.com/questions/27278222/is-it-possible-to-add-a-plugin-to-chromedriver-under-a-protractor-test/27278322#27278322

Apparently, Selenium WebDriver is expecting a Base64 encoded string of the crx file, not the path to the file. This will take some time to fix and test. 

I was able to successfully test one of the alternative answers in that stackoverflow link, in case it is of use to you until I can fix the `AddExtensions` problem:
```vba
objWebCapabilities.AddArguments "--load-extension=" & Environ("LOCALAPPDATA") & "\Google\Chrome\User Data\Default\Extensions\jdbnofccmhefkmjbkkdkfiicjkgofkdh\2.3.1_0"
```

**Comment by GCuser99** [comment]
@dornech, before I commit the fix, am hoping you can test for me....
```vba
Sub TestExtensions()
    Dim objSelenium As New WebDriver
    Dim objWebCapabilities As WebCapabilities
    Dim b64 As String
    Dim strExtensionPath As String

    strExtensionPath = "[path to local crx file]"

    b64 = encodeBase64(readByteArrayFromFile(strExtensionPath))
    
    objSelenium.StartChrome
    
    Set objWebCapabilities = objSelenium.CreateCapabilities
    objWebCapabilities.AddExtensions b64
    
    objSelenium.OpenBrowser objWebCapabilities
    
    Stop 'check if extension loaded
    
    objSelenium.CloseBrowser
    objSelenium.Shutdown
End Sub

Private Function encodeBase64(bytes() As Byte) As String
    Dim domDoc As New MSXML2.DOMDocument60
    Dim domNode As MSXML2.IXMLDOMElement
    Set domNode = domDoc.createElement("b64")
    domNode.DataType = "bin.base64"
    domNode.nodeTypedValue = bytes
    encodeBase64 = domNode.text
End Function

Private Function readByteArrayFromFile(ByVal filePath As String) As Byte()
    Dim binaryStream As New ADODB.Stream
    Dim byteArray() As Byte

    'specify stream type - we want To save binary data.
    binaryStream.Type = adTypeBinary
  
    'open the stream
    binaryStream.Open
    
    'load and read binary data from disk
    binaryStream.LoadFromFile filePath
    byteArray = binaryStream.Read
  
    binaryStream.Close
    readByteArrayFromFile = byteArray
End Function
```

**Comment by dornech** [comment]
Hi, that's funny: based on your hint I was looking for a possibility to do the b64 encoding - and found the MSXML-thing. And I have already a read2bytearray routine in my VBA pool. Just tested the bits & pieces togehter and wanted to suggest this to you ...

**Comment by dornech** [comment]
Slightly different variable names etc. but in principle yes.

3 Points to note: 
- For upward compatibility I wonder if there is a posibility to check if data provided is already a base64 string i e avoid code changes and do the encoding only if a path/filename is provided. 
- I wonder if this is a browser dependent issue - I tested only Chrome (and Chromium style browsers/drivers) and I am not sure nor did I investigate if it applies to Firefox as well.
-  I ran into a memory issue with one extension (i. e. AdblockPlus). I do not know if this can be avoided or is a restriction of the XML library.

**Comment by GCuser99** [comment]
Oops, I made the commit literally seconds before your last post, so missed your points above in the update.

Anyway, thanks for reporting this and let us know how the transition to SeleniumVBA goes.

Mike

**Comment by GCuser99** [comment]
BTW, I tried with AdBlockPlus on Windows 11, latest Chrome browser. The crx file is large (64 mb) compared to the other crx files that I tried (less than a 1 mb), but the encoding worked very fast (less than a second). The sending of the capabilities instruction with the encoded string to the Chrome WebDriver and waiting for the response to come back took about 20 seconds or so. It did seem to work - AdBlockPlus presented a page asking me to pay money. So I don't think it's a problem with the encoding, at least on my setup.
